contracts/ServiceNodeContribution.sol

```diff
@@ -230,72 +176,88 @@ contract ServiceNodeContribution is Shared {
+
+            if (reservedContributions[reserved[i].addr] != 0)
+                revert DuplicateAddressInReservedContributor(i);
 
             // NOTE: Check contribution meets min requirements and running sum
             // doesn't exceed a full stake
             uint256 minContrib     = calcMinimumContribution(remaining, i, maxContributors);
-            uint256 contribAmount  = reserved[i].stakedAmount;
-            require(contribAmount >= minContrib, "Contribution is below minimum requirement");
-            require(remaining >= contribAmount,  "Sum of reserved contribution slots exceeds the staking requirement");
-            remaining         -= contribAmount;
+            uint256 contribAmount  = reserved[i].amount;
+
+            if (contribAmount < minContrib)
+                revert ReservedContributionBelowMinAmount(i, contribAmount, minContrib);
+
+            if (remaining < contribAmount)
+                revert ReservedContributionExceedsStakingRequirement(i, contribAmount, remaining);
+
+            remaining -= contribAmount;
 
             // NOTE: Store the reservation in the contract
             reservedContributionsAddresses.push(reserved[i].addr);
             reservedContributions[reserved[i].addr] = contribAmount;
         }
     }
 
-    /// @notice Contribute funds to the contract for the node run by
-    /// `operator`. The `amount` of SENT token must be at least the
-    /// `minimumContribution` or their amount specified in their reserved
-    /// contribution (if applicable) otherwise the contribution is reverted.
-    ///
-    /// Node registration parameters must be assigned prior to the operator
-    /// contribution or alternatively after `reset` is invoked.
-    ///
-    /// The operator must contribute their minimum contribution/reservation
-    /// before the public or reserved contributors can contribute to the node.
-    /// The minimum an operator can contribute is 25% of the staking requirement
-    /// regardless of having a reservation or not.
-    ///
-    /// @param amount The amount of SENT token to contribute to the contract.
-    function contributeFunds(uint256 amount) external { _contributeFunds(msg.sender, amount); }
+    function updateBeneficiary(address newBeneficiary) external { _updateBeneficiary(msg.sender, newBeneficiary); }
+
+    function _updateBeneficiary(address stakerAddr, address newBeneficiary) private {
+        bool updated   = false;
+        uint256 length = _contributorAddresses.length;
+        for (uint256 i = 0; i < length; i++) {
+            IServiceNodeRewards.Staker storage staker = _contributorAddresses[i];
+            if (staker.addr == stakerAddr) {
+                updated            = true;
+                staker.beneficiary = newBeneficiary;
+                break;
+            }
+        }
+
+        if (!updated)
+            revert NonContributorUpdatedBeneficiary(stakerAddr);
+
+        emit UpdateStakerBeneficiary(stakerAddr, newBeneficiary);
```

Comment:
Do you think it's worthwhile adding the old beneficiary address in here as well (so that event log watchers can show a "x updated beneficiary address from y to z" instead of just "x updated beneficiary address to z")?

Reply:
Yep. That'd be useful, the thing I only realised about events just recently is that the event it emits should contain all the information related to the event because that makes it more useful to be handled atomically rather than require the caller to re-query the contract or have to cache the state prior to the event being logged.